clients/rabbitmq/src/main/scala/mqperf/rabbitmq/RabbitMq.scala

```diff
@@ -0,0 +1,200 @@
+
+    cf.useNio()
+    cf.setNioParams(new NioParams().setNbIoThreads(50))
+
+    try {
+      conn = cf.newConnection()
+
+      newChannel(queueName, passive = false)
+      logger.info(s"Created queue $queueName")
+    } catch {
+      case e: Exception => logger.warn(s"Queue $queueName creation failed", e)
+    }
+  }
+
+
+  override def cleanUp(config: Config): Unit = {
+    logger.info("Closing connection")
+    conn.close()
+  }
+
+  private def newChannel(queueName: String, passive: Boolean): Channel = {
+    val channel = conn.createChannel()
+    if (passive) channel.queueDeclarePassive(queueName)
+    else channel.queueDeclare(queueName, false, false, false, null)
+
+    channel
+  }
+
+  override def createSender(config: Config): MqSender = new MqSender {
+    val channelPoolList: CopyOnWriteArrayList[(Channel, Int)] =
+      new CopyOnWriteArrayList[(Channel, Int)]()
+    val maxChannelsNr: Int = Option(config.mqConfig(MaxChannelsNrConfigKey)).map(_.toInt).getOrElse(1)
+    val maxSendInFlight: Int = config.maxSendInFlight
+    val batchSizeSend: Int = config.batchSizeSend
+    private val queueName: String = config.mqConfig(QueueNameConfigKey)
+
+    for (_ <- 1 to maxChannelsNr) {
+      val channel = conn.createChannel()
+      channel.queueDeclarePassive(queueName)
+      channel.confirmSelect()
+      channelPoolList.add((channel, maxSendInFlight))
+    }
+
+    /*
+      Send method uses channels pool in order to be able to send messages using the sender instance on multiple threads simultaneously.
+      Channel has a limit of messages that it can publish without ack - maxInFlight TODO: mstopyra (Question? - is it global - shared for channels)
+      After publishing messages the limit for the channel decreases (minus size of a send batch)
+      Releasing a channel (adding it back to the pool) happens after publishing a certain batch, according to the following logic:
+       1. if the limit that is left is smaller than the size of a send batch (also 0)
+          - the channel is waiting for send ack and after that it is appended to the channel pool and the limit gets reset to the init value
+       2. if the limit that is left is greater or equal the size of a send batch
+          - the channel gets prepended to the list with the decreased limit
+     */
+    override def send(msgs: Seq[String]): Future[Unit] = {
+      pollChannelUntilPresent()
+        .flatMap { case (channel, leftInFlightSizeBeforePublish) =>
+          Future {
+            blocking {
+              msgs.foreach(msg => channel.basicPublish("", queueName, MessageProperties.PERSISTENT_TEXT_PLAIN, msg.getBytes))
+              logger.info(s"publishing message - channel ${channel.toString}")
+              val leftInFlightSizeAfterPublish = leftInFlightSizeBeforePublish - msgs.size
+              if (leftInFlightSizeAfterPublish >= batchSizeSend) {
+                channelPoolList.add(0, (channel, leftInFlightSizeAfterPublish))
```

Comment:
a `CopyOnWriteArray` should only ever be updated from a single thread (but it can be read from many threads); again, I think that's the wrong data structure to use here